linalg/dual.hpp

```diff
@@ -0,0 +1,413 @@
+{
+   return {a.value - b.value, a.gradient - b.gradient};
+}
+
+/** @brief multiplication of a dual number and a non-dual number */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+constexpr auto operator*(const dual<value_type, gradient_type>& a,
+                         double b) -> dual<decltype(a.value * b), decltype(a.gradient * b)>
+{
+   return {a.value * b, a.gradient * b};
+}
+
+/** @brief multiplication of a dual number and a non-dual number */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+constexpr auto operator*(double a,
+                         const dual<value_type, gradient_type>& b) ->
+dual<decltype(a * b.value), decltype(a * b.gradient)>
+{
+   return {a * b.value, a * b.gradient};
+}
+
+/** @brief multiplication of two dual numbers */
+template <typename value_type_a, typename gradient_type_a, typename value_type_b, typename gradient_type_b>
+MFEM_HOST_DEVICE
+constexpr auto operator*(dual<value_type_a, gradient_type_a> a,
+                         dual<value_type_b, gradient_type_b> b) -> dual<decltype(a.value * b.value),
+                              decltype(b.value * a.gradient + a.value * b.gradient)>
+{
+   return {a.value * b.value, b.value * a.gradient + a.value * b.gradient};
+}
+
+/** @brief division of a dual number by a non-dual number */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+constexpr auto operator/(const dual<value_type, gradient_type>& a,
+                         double b) -> dual<decltype(a.value / b), decltype(a.gradient / b)>
+{
+   return {a.value / b, a.gradient / b};
+}
+
+/** @brief division of a non-dual number by a dual number */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+constexpr auto operator/(double a,
+                         const dual<value_type, gradient_type>& b) -> dual<decltype(a / b.value),
+                               decltype(-(a / (b.value * b.value)) * b.gradient)>
+{
+   return {a / b.value, -(a / (b.value * b.value)) * b.gradient};
+}
+
+/** @brief division of two dual numbers */
+template <typename value_type_a, typename gradient_type_a, typename value_type_b, typename gradient_type_b>
+MFEM_HOST_DEVICE
+constexpr auto operator/(dual<value_type_a, gradient_type_a> a,
+                         dual<value_type_b, gradient_type_b> b) -> dual<decltype(a.value / b.value),
+                              decltype((a.gradient / b.value) -
+                                       (a.value * b.gradient) /
+                                       (b.value * b.value))>
+{
+   return {a.value / b.value, (a.gradient / b.value) - (a.value * b.gradient) / (b.value * b.value)};
+}
+
+/**
+ * @brief Generates const + non-const overloads for a binary comparison operator
+ * Comparisons are conducted against the "value" part of the dual number
+ * @param[in] x The comparison operator to overload
+ */
+#define binary_comparator_overload(x)                           \
```

Comment:
Since this is in a public header, it is better to add the prefix `MFEM_` to this macro. Even if it is `#undef`ined below.